### PR TITLE
feat: Open up AbstractMultiSelectionModel for extension.

### PR DESCRIPTION
**Pull request by OlliTietavainenVaadin**
## Description

It's not possible to implement a custom lazy multi-selection model that doesn't fetch the entire dataset.

Open up the AbstractGridMultiSelectionModel API a bit so extension becomes possible: make clientSelectAll protected instead of package-protected and add a protected getter for the selectionColumn.  Part of #4780.

## Type of change

- [ ] Bugfix
- [x ] Feature

## Checklist

- [x ] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x ] I have added a description following the guideline.
- [x ] The issue is created in the corresponding repository and I have referenced it.
- [x ] I have added tests to ensure my change is effective and works as intended.
- [x ] New and existing tests are passing locally with my change.
- [x ] I have performed self-review and corrected misspellings.
- [] I have not completed some of the steps above and my pull request can be closed immediately.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java [hunk 1]
```diff
@@ -365,7 +365,7 @@ public boolean isSelectionColumnFrozen() {
     protected abstract void fireSelectionEvent(
             SelectionEvent<Grid<T>, T> event);
 
-    void clientSelectAll() {
+    protected void clientSelectAll() {
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:368]: We might as well expose `clientDeselectAll` for consistency.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java
[the remark this answers is not in the document]
Comment (OlliTietavainenVaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:1]: Example here: https://github.com/OlliTietavainenVaadin/skeleton-starter-flow-24/blob/main/src/main/java/org/vaadin/example/MainView.java

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/customselectionmodel/CustomSelectionModelTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package com.vaadin.flow.component.grid.customselectionmodel;
+
+import com.vaadin.flow.component.ComponentEvent;
+import com.vaadin.flow.component.grid.AbstractGridMultiSelectionModel;
+import com.vaadin.flow.component.grid.Grid;
+import com.vaadin.flow.data.selection.SelectionEvent;
+import org.junit.Test;
+
+public class CustomSelectionModelTest {
```
[the remark this answers is not in the document]
Comment (OlliTietavainenVaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/customselectionmodel/CustomSelectionModelTest.java:9]: Removed

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java (hunk 1 above)
Comment (OlliTietavainenVaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:368]: Done

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java
[the remark this answers is not in the document]
Comment (sissbruecker) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:1]: You would also need to override the single item select methods to keep the select all checkbox in a consistent state. For example:
- Select all
- Deselect item
- Select item again

Now the select all checkbox is in indeterminate state. A proper implementation will probably require bookkeeping of the items that have been deselected after a select all operation, as well as custom logic for determining the actual list of selected items. Still, it looks like all selection methods can be overriden, and there are also accessible methods for manipulating the internal selection state, so we can probably assume that a custom implementation should be possible now.

**Review by sissbruecker — APPROVED** [review]

**Pull request merged**